query/query/query.py

```diff
@@ -52,107 +62,92 @@ async def healthcheck(request):  # pylint: disable=unused-argument
 
 
-def blocking_execute(jbackend, username, session_id, billing_project, bucket, code):
-    return jbackend.execute(username, session_id, billing_project, bucket, code)
+def blocking_execute(jbackend, userdata, body):
+    return jbackend.execute(userdata['username'], userdata['session_id'], body['billing_project'], body['bucket'], body['code'])
 
 
-@routes.post('/execute')
-@rest_authenticated_users_only
-async def execute(request, userdata):
-    app = request.app
-    thread_pool = app['thread_pool']
-    jbackend = app['jbackend']
-    body = await request.json()
-    billing_project = body['billing_project']
-    bucket = body['bucket']
-    code = body['code']
-    log.info(f'execute: {code}')
-    await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_execute, jbackend, userdata['username'], userdata['session_id'], billing_project, bucket, code)
-    return java_to_web_response(jresp)
+def blocking_value_type(jbackend, userdata, body):
+    return jbackend.valueType(userdata['username'], body['code'])
 
 
-def blocking_value_type(jbackend, username, code):
-    return jbackend.valueType(username, code)
+def blocking_table_type(jbackend, userdata, body):
+    return jbackend.tableType(userdata['username'], body['code'])
 
 
-@routes.post('/type/value')
-@rest_authenticated_users_only
-async def value_type(request, userdata):
-    app = request.app
-    thread_pool = app['thread_pool']
-    jbackend = app['jbackend']
-    code = await request.json()
-    log.info(f'value type: {code}')
-    await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_value_type, jbackend, userdata['username'], code)
-    return java_to_web_response(jresp)
+def blocking_matrix_type(jbackend, userdata, body):
+    return jbackend.matrixTableType(userdata['username'], body['code'])
 
 
-def blocking_table_type(jbackend, username, code):
-    return jbackend.tableType(username, code)
+def blocking_blockmatrix_type(jbackend, userdata, body):
+    return jbackend.blockMatrixType(userdata['username'], body['code'])
 
 
-@routes.post('/type/table')
-@rest_authenticated_users_only
-async def table_type(request, userdata):
+def blocking_get_reference(jbackend, userdata, body):   # pylint: disable=unused-argument
+    return jbackend.referenceGenome(userdata['username'], body['name'])
+
+
+async def handle_ws_response(request, userdata, endpoint, f):
     app = request.app
-    thread_pool = app['thread_pool']
     jbackend = app['jbackend']
-    code = await request.json()
-    log.info(f'table type: {code}')
+
     await add_user(app, userdata)
-    jresp = await blocking_to_async(thread_pool, blocking_table_type, jbackend, userdata['username'], code)
-    return java_to_web_response(jresp)
+    log.info(f'{endpoint}: connecting websocket')
+    ws = web.WebSocketResponse(heartbeat=30, max_msg_size=0)
+    task = None
+    try:
+        await ws.prepare(request)
+        app['sockets'].add(ws)
```

Comment:
I still assert this is unnecessary. This web socket is always owned by a coroutine owned by aiohttp. Aiohttp will cancel the handler if the server is shutting down or the request is terminated. When this coroutine receives a TimeoutError, it will hit the finally block and kill the web socket. The conversation in "Graceful Shutdown" applies to web sockets owned by tasks that aren't connected to aiohttp by an ownership chain.

Reply:
I did take the rest of it out, but somehow missed this line…